ludwig/data/preprocessing.py

```diff
@@ -600,6 +503,172 @@ def preprocess_for_training(
+            data,
+            data['split']
+        )
+        if not skip_save_processed_input:
+            logging.info('Writing dataset')
+            data_train_hdf5_fp = get_filepath_in_diff_format(data_train_csv,
+                                                             'hdf5')
+            data_utils.save_hdf5(
+                data_train_hdf5_fp,
+                training_set,
+                train_set_metadata
+            )
+            if validation_set is not None:
+                data_validation_hdf5_fp = get_filepath_in_diff_format(
+                    data_validation_csv,
+                    'hdf5'
+                )
+                data_utils.save_hdf5(
+                    data_validation_hdf5_fp,
+                    validation_set,
+                    train_set_metadata
+                )
+            if test_set is not None:
+                data_test_hdf5_fp = get_filepath_in_diff_format(data_test_csv,
+                                                                'hdf5')
+                data_utils.save_hdf5(
+                    data_test_hdf5_fp,
+                    test_set,
+                    train_set_metadata
+                )
+            logging.info('Writing train set metadata with vocabulary')
+            train_set_metadata_json_fp = get_filepath_in_diff_format(data_csv,
+                                                                     'json')
+            data_utils.save_json(
+                train_set_metadata_json_fp, train_set_metadata)
+
+    return training_set, test_set, validation_set, train_set_metadata
+
+
+def _preprocess_df_for_training(
+        features,
+        data_df=None,
+        data_train_df=None,
+        data_validation_df=None,
+        data_test_df=None,
+        preprocessing_params=default_preprocessing_parameters,
+        random_seed=default_random_seed
+):
+    """ Method to pre-process dataframes. This doesn't have the optoin to save the
+    processed data as hdf5 as we don't expect users to do this as the data can
+    be processed in memory
+    """
+
+    if data_df is not None:
+        # needs preprocessing
+        logging.info('Using full dataframe')
+        logging.info('Building dataset (it may take a while)')
+        data, train_set_metadata = build_dataset_df(
+            data_df,
+            features,
+            preprocessing_params,
+            random_seed=random_seed
+        )
+        training_set, test_set, validation_set = split_dataset_tvt(
+            data,
+            data['split']
+        )
+
+    elif data_train_df is not None:
+        # needs preprocessing
+        logging.info('Using training dataframe')
+        logging.info('Building dataset (it may take a while)')
+        concatenated_df = concatenate_df(
+            data_train_df,
+            data_validation_df,
+            data_test_df
+        )
+        data, train_set_metadata = build_dataset_df(
```

Comment:
Is there a way to combine this call with the one in line 638?

For example first get a combined dataframe: df_combined = data_df (if data_df is not None) or df_combiner is a concatenated dataframe if data_train_df is not none, etc.

And then call build_dataset_df on that dataframe.

Reply:
yes! missed that. will do now